src/digests.c

```diff
@@ -0,0 +1,492 @@
+        dctx->session = CK_INVALID_HANDLE; \
+        return dctx; \
+    }
+
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA_1, sha1);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA224, sha224);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA256, sha256);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA384, sha384);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA512, sha512);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA512_224, sha512_224);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA512_256, sha512_256);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA3_224, sha3_224);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA3_256, sha3_256);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA3_384, sha3_384);
+DISPATCH_DIGEST_NEWCTX_FN(CKM_SHA3_512, sha3_512);
+DISPATCH_DIGEST_COMMON_FN(dupctx);
+DISPATCH_DIGEST_COMMON_FN(freectx);
+DISPATCH_DIGEST_COMMON_FN(init);
+DISPATCH_DIGEST_COMMON_FN(update);
+DISPATCH_DIGEST_COMMON_FN(final);
+DISPATCH_DIGEST_COMMON_FN(gettable_params);
+
+static void *p11prov_digest_dupctx(void *ctx)
+{
+    P11PROV_DIGEST_CTX *dctx = (P11PROV_DIGEST_CTX *)ctx;
+    P11PROV_DIGEST_CTX *newctx;
+    CK_SLOT_ID slotid = CK_UNAVAILABLE_INFORMATION;
+    CK_SESSION_HANDLE sess = CK_INVALID_HANDLE;
+    CK_FUNCTION_LIST *f;
+    CK_BYTE_PTR state = NULL;
+    CK_ULONG state_len;
+    CK_RV ret;
+
+    P11PROV_debug("digest dupctx, ctx=%p", ctx);
+
+    if (dctx == NULL) {
+        return NULL;
+    }
+
+    newctx = OPENSSL_zalloc(sizeof(P11PROV_DIGEST_CTX));
+    if (dctx == NULL) {
+        return NULL;
+    }
+    newctx->provctx = dctx->provctx;
+    newctx->mechtype = dctx->mechtype;
+
+    if (dctx->session == NULL) {
+        goto done;
+    }
+
+    /* This is not really funny. OpenSSL by default assumes contexts with
+     * operations in flight can be easily duplicated, with all the
+     * cryptographic status and then both context can keep going
+     * independently. We'll try here, but on failure we just 'move' the
+     * to the new token and hope for the best */
+
+    ret = p11prov_ctx_status(dctx->provctx, &f);
+    if (ret != CKR_OK) {
+        return NULL;
+    }
+
+    sess = p11prov_session_handle(dctx->session);
+
+    /* move old session to new context, we swap because often openssl continues
+     * on the duplicated context by default */
+    newctx->session = dctx->session;
+    dctx->session = NULL;
+
+    /* NOTE: most tokens will probably return errors trying to do this on digest
+     * sessions */
+
+    ret = f->C_GetOperationState(sess, NULL_PTR, &state_len);
+    if (ret != CKR_OK) {
+        P11PROV_raise(dctx->provctx, ret,
+                      "Error returned by C_GetOperationState");
+        goto done;
+    }
+    state = OPENSSL_malloc(state_len);
+    if (state == NULL) {
+        goto done;
+    }
+
+    ret = f->C_GetOperationState(sess, state, &state_len);
+    if (ret != CKR_OK) {
+        P11PROV_raise(dctx->provctx, ret,
+                      "Error returned by C_GetOperationState");
+        goto done;
+    }
+
+    /* FIXME: should probably cycle trhough slots and check if the slot,
```

Comment:
Typo: `s/trhough/through/`

Reply:
I need to find a better spell checker, the one in the CI doesn't catch much ...

Reply:
fixed